core-bundle/src/Resources/contao/library/Contao/System.php

```diff
@@ -257,13 +276,42 @@ public static function importStatic($strClass, $strKey=null, $blnForce=false)
 			}
 			else
 			{
-				static::$arrStaticObjects[$strKey] = new $strClass();
+				try
+				{
+					static::$arrStaticObjects[$strKey] = new $strClass();
+				}
+				catch (\Throwable $t)
+				{
+					if (!$container->getParameter('kernel.debug') && self::isServiceInlined($strClass))
+					{
+						throw new ServiceNotFoundException($strClass, null, null, array(), sprintf('The "%s" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.', $strClass));
+					}
+
+					throw $t;
+				}
 			}
 		}
 
 		return static::$arrStaticObjects[$strKey];
 	}
 
+	private static function isServiceInlined($strClass)
+	{
+		$container = static::getContainer();
+
+		if (!$container instanceof Container)
+		{
+			return false;
+		}
+
+		if (null === static::$removedServiceIds)
+		{
+			static::$removedServiceIds = $container->getRemovedIds();
+		}
+
+		return isset(static::$removedServiceIds[$strClass]);
```

Comment:
As `$removedServiceIds` is `private`, shouldn’t it be accessed via `self`?

```suggestion
		if (null === self::$removedServiceIds)
		{
			self::$removedServiceIds = $container->getRemovedIds();
		}

		return isset(self::$removedServiceIds[$strClass]);
```